### ISSUE TITLE
low coverage error and warning regarding methylation tags in v1.9.1

### ISSUE THREAD

**Issue by Fazulur** [epi2me-labs/wf-human-variation#129]
Dear Team,

I tried v1.9.1 version of the pipeline on our data and for some of the samples, though methylation tags are available in unmapped input bam, pipeline giving below warning.

**WARN: Alignment files do not contain modified base tags. Skipping mod aggregation and reporting.**

Here is the one line from unmapped bam showing methylation tags.

b31b37f0-e31c-4f27-9d47-41bdd1856037	4	*	0	0	*	*	0	0	ATGTTTCTGCCTACATTCATACATTGCTACCTCCAGTTGACTTCCTTTTTCACTCTTTCCATACCAATATCCTCTTCCTTTGTCCCTGTTTCAATTACTTTCCCTCAAATTTAGCATCTATCTCAGCTACACATTAGAAACTTCTTGAGAGTGGGTGGTGAGTCAGAGAGGGCTAAATATTCCAACACCCAGGCCACACTCCAGGCCAATAAAAGTAGAATCTGGGGGGTGAAGCTCTTCGTGGGCAATTTTAAAGGTCCATCTTTCTGGG	()*+,*&''(%%%$#$&%%'$#$%%'''%&(*'((+/.))))/*/30404/+**&%&++'&&$$&$*/)'&&'&&(&3432)6659965468854523324....997787775567665456798754557555688888789:=>;9878:999989555554445563.*0-+++**))+/145887698798:888::::88:;<5445:85.+++++,(%$$$$$%$'&&&%&(%$$$%%&(*)*+//344556760+'######%	MM:Z:C+h?,63;C+m?,63;	**ML:B:C,4,2**	mv:B:c,6,1,0,0,1,0,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,1,0,1,0,0,0,1,0,1,0,0,0,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0,0,0,0,0,1,1,1,1,0,1,1,0,0,1,1,1,0,1,1,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,0,1,1,0,0,0,0,0,0,1,1,0,0,0,1,0,1,1,0,0,1,1,1,1,0,1,1,0,1,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,1,0,1,1,0,1,1,1,1,0,1,1,0,1,1,0,1,0,0,0,0,1,1,1,0,1,1,1,1,0,1,0,1,1,0,0,0,0,1,0,1,1,0,1,1,0,1,0,1,0,1,0,0,0,0,0,1,0,1,0,1,1,0,1,1,1,0,1,1,1,0,1,1,0,1,1,0,0,1,0,1,1,1,0,1,1,1,0,1,0,1,1,0,1,0,1,1,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,1,0,0,0,0,0,0,1,0,1,0,0,0,0,1,1,0,1,1,0,1,1,1,1,0,1,1,1,1,0,1,1,1,0,1,0,1,1,0,0,1,0,1,0,0,1,1,1,0,1,1,1,1,0,1,1,0,1,1,0,1,0,1,0,0,0,1,0,1,1,0,0,1,0,1,0,1,0,1,1,1,0,1,1,1,0,1,1,1,0,1,1,0,1,1,1,0,1,1,1,1,0,1,1,1,1,0,1,1,0,0,0,0,0,1,1,0,1,0,1,1,1,0,1,0,1,1,0,1,0,1,1,0,1,1,0,1,1,0,1,1,1,1,0,1,1,0,1,1,0,1,1,0,1,0,1,1,1,0,1,0,1,1,0,0,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,0,1,0,1,1,1,0,0,1,0,1,0,0,1,0,1,0,0,0,1,1,1,0,0,0,0,1,1,0,1,1,0,1,1,1,0,1,1,1,0,1,0,1,1,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,1,0,0,1,0,1,0,0	qs:i:11	ns:i:3313	ts:i:10	sm:f:99.8018	sd:f:27.3646	sv:Z:quantile	qs:i:11	mx:i:4	ch:i:791	rn:i:182	st:Z:2023-11-29T08:03:04.580025+03:00	ns:i:3313	ts:i:10	sm:f:99.8018	sd:f:27.3646	sv:Z:med_mad	RG:Z:dc45353d-4fd3-4505-b0d4-d36015eb53c6_dna_r10.4.1_e8.2_400bps_hac@v4.2.0

In addition to above warning, pipeline terminated with low coverage error. I assume this is default behavior of the pipeline where it will accept only >20X coverage. Is there any option to continue process the low coverage samples in these cases?

**ERROR ~ ERROR: File test.merged.bam will not be processed by the workflow as the detected coverage of 17.552x is below the minimum coverage threshold of 20x required for analysis.**

In addition to this, in some samples, although the run is completed without any issues, the trace file indicates FAILED status tasks. The process attempts multiple times, and the last attempt is successful.

In the example below, the 'snp:pileup_variants (174)' step ran multiple times. The first attempt failed and still appears in the trace file, while the second attempt shows as completed. This gave me the impression that the pipeline didn't finish, as the trace file contains some tasks with FAILED status. Is this the default behavior?

210	b0/c143ce	769768	snp:pileup_variants (174)	**FAILED**	143	2023-12-10 23:23:17.749	3m 55s	3m 50s	-	-	-	-	-
669	23/db9ff6	773317	snp:pileup_variants (174)	**COMPLETED**	0	2023-12-10 23:48:32.989	7m 25s	7m 21s	98.7%	3.2 GB	8.3 GB	1.5 GB	80.1 MB

Please help me to resolve this.

Thanks In advance
Fazulur Rehaman

**Comment by RenzoTale88** [comment]
@Fazulur in order:
1. Is the BAM file a combination of multiple, individual BAMs? Have all the reads the MM/ML tags, or just some of them?
2. you can change the BAM coverage threshold with `--bam_min_coverage N`
3. It can happen that some indivudal tasks fail at runtime. These are picked up by nextflow and retried. In this case, the first is ignored, so if subsequent attempt succeed there is nothing to worry about. 

Andrea

**Comment by Fazulur** [comment]
Hi Andrea,

Thank you so much. y

yes bam is combination of mulitple individual bams, I have merged them to one bam. And all reads have MM/ML tags.

And, thanks for the option to proceed the  analysis with low coverage data. 

Thanks & Regards
Fazulur Rehaman

**Comment by RenzoTale88** [comment]
@Fazulur do you still come across the lack of MM/ML tag warning?

**Comment by Fazulur** [comment]
Hi Andrea,

Yes, I am still getting MM/ML tag warning with v.1.9.1.

Thanks & Regards
Fazulur Rehaman

**Comment by RenzoTale88** [comment]
@Fazulur out script currently checks the first 1000 entries of the BAM for the presence of at least one read with both the MM and ML tags. Does your BAM have 10K reads at the beginning with both MM and ML tags? Sorry to check this once more, but it might help us narrow down what the issue is.

**Comment by Fazulur** [comment]
Hi Andrea,

Yes, i can see MM/ML tags in first 1000 entries of BAM.

Here is the output of modkit summary

**> sampling 10000 reads from BAM
> calculating threshold at 10% percentile
> calculated thresholds: C: 0.7519531
# bases             C 
# total_reads_used  10000 
# count_reads_C     10000 
# pass_threshold_C  0.7519531 
 base  code  pass_count  pass_frac    all_count  all_frac 
 C     h     19049       0.034969874  39459      0.06523626 
 C     m     407799      0.7486314    441123     0.72929406 
 C     -     117878      0.2163987    124281     0.20546967** 

Thanks & Regards
Fazulur Rehaman

**Comment by RenzoTale88** [comment]
Thanks for confirming this. Do you mind sharing the `.nextflow.log` file?

**Comment by Fazulur** [comment]
Hi Andrea,

Please find the attached .nextflow.log.
[test.nextflow.log](https://github.com/epi2me-labs/wf-human-variation/files/13921730/test.nextflow.log)

Thanks & Regards
Fazulur Rehaman

**Comment by RenzoTale88** [comment]
@Fazulur I had a look at the log file and it appears that the workflow completed the modkit_phase process successfully:
```
grep 'modkit' test.nextflow.log | grep status | tail -1
Dec-21 21:25:37.517 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[jobId: 976829; id: 1401; name: mod:modkit_phase (1); status: COMPLETED; exit: 0; error: -; workDir: ONT/20231003/test/78/b6c5b23ea69a4616aa1516585e12bc started: 1703188437356; exited: 2023-12-21T21:25:34.309699Z; ]
```

And the log message doesn't appear to be in the log:
```
grep 'Alignment files do not contain modified base tags.' test.nextflow.log | wc -l
0
```

Do you see the outputs from `modkit` in the output directory? Is it the right log?

**Comment by Fazulur** [comment]
Hi Andrea,

Yes, this is correct log. I am using lsf as executor and i can see WARNING message only in standard error log of LSF and not in .nextflow.log. 

results (test.merged_1.wf_mods.bedmethyl.gz, test.merged_2.wf_mods.bedmethyl.gz & test.merged_ungrouped.wf_mods.bedmethyl.gz) are generated in output directory.  

Thanks & Regards
Fazulur Rehaman

**Comment by RenzoTale88** [comment]
Hi @Fazulur the nextflow.log should report also the error/warnings, as you can see in line 
```
grep 'WARN' test.nextflow.log
grep 'WARN' /mnt/c/Users/atalenti/Downloads/test.nextflow.log
Dec-21 09:29:36.021 [main] WARN  nextflow.Nextflow - CNV calling subworkflow does not support CRAM. You don't need to do anything, but we're just letting you know that:
Dec-21 09:29:36.022 [main] WARN  nextflow.Nextflow - - If your input file is CRAM, it will be converted to a temporary BAM inside the workflow automatically.
Dec-21 09:29:36.022 [main] WARN  nextflow.Nextflow - - If your input requires alignment or basecalling, the outputs will be saved to your output directory as BAM instead of CRAM.
```

Moreover, when the workflow display that message, it should skip `modkit`. The fact that you are seeing the results makes me think that the workflow has completed successfully. I apologize, but I'm unsure how I can be of help as the modkit workflow completed successfully.

**Comment by Fazulur** [comment]
Hi Andrea,

Thank you. I apologize for not providing you correct testcase. I have observed in other samples that we get only this warning when we have coverage threshold < 20x.

Thanks & Regards
Fazulur Rehaman

**Comment by Fazulur** [comment]
Hi Andrea,

Thank you so much.

Thanks & Regards
Fazulur Rehaman

**Comment by SamStudio8** [comment]
Fixed in 1.10.1, thanks for your report!
